Reads and writes are propagated differently from unreachable code in local functions and behavior is not consistent across different types of unrachable code.

In the following code the call to 'Local' should either propagate both the write to 'x' and the read from 'y' or neither of them or mark all code after the call as unreachable.
Currently the call propagates write to 'x' though it is unreachable but doesn't propagate unreachable read from 'y'.

```cs
class C
{
  public void M3()
  {
    int x, y;
    void Local()
    {
      while(true) continue;
      x = 5; // unreachable code
      System.Console.WriteLine(y);
    }

    Local(); // the call doesn't require 'y' to be assigned since its usage is unreachable
    System.Console.WriteLine(x); // 'x' is assigned here but it shouldn't since assignment is unreachable

    // the error here implies the branch itself is reachable
    System.Console.WriteLine(y); // error CS0165: Use of unassigned local variable 'y'
  }
}
```

unreachable code after a 'throw' statement behaves the same way.

```cs
class C
{
  public void M3()
  {
    int x, y;
    void Local()
    {
      throw null;
      x = 5; // unreachable code
      System.Console.WriteLine(y);
    }

    Local(); // the call doesn't require 'y' to be assigned since its usage is unreachable
    System.Console.WriteLine(x); // 'x' is assigned here but it shouldn't since assignment is unreachable

    // the error here implies the branch itself is reachable
    System.Console.WriteLine(y); // error CS0165: Use of unassigned local variable 'y'
  }
}
```

Unreachable code after 'return' statement behaves differently and usages there don't affect compilation.

```cs
class C
{
  public void M3()
  {
    int x, y;
    void Local()
    {
      return;
      x = 5;
      System.Console.WriteLine(y);
    }

    Local();
    System.Console.WriteLine(x); // error CS0165: Use of unassigned local variable 'x'
    System.Console.WriteLine(y);
  }
}
```

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.LocalFunctions.cs
@@ -16,7 +16,14 @@ namespace Microsoft.CodeAnalysis.CSharp
         private class LocalFuncUsages
         {
             public BitVector ReadVars = BitVector.Empty;
-            public BitVector WrittenVars = BitVector.Empty;
+            public LocalState WrittenVars;
+
+            public LocalFuncUsages(LocalState unreachableState)
+            {
+                // If we have yet to analyze the local function
+                // definition, assume it definitely assigns everything
+                WrittenVars = unreachableState;
+            }
 
             public bool LocalFuncVisited { get; set; } = false;
         }
@@ -32,45 +39,31 @@ namespace Microsoft.CodeAnalysis.CSharp
         {
             _usedLocalFunctions.Add(localFunc);
 
+            var usages = GetOrCreateLocalFuncUsages(localFunc);
+
             // First process the reads
-            ReplayVarUsage(localFunc,
-                           syntax,
-                           isWrite: false);
+            ReplayReads(ref usages.ReadVars, syntax);
 
             // Now the writes
             if (writes)
             {
-                ReplayVarUsage(localFunc,
-                               syntax,
-                               isWrite: true);
-            }
-        }
-
-        private void ReplayVarUsage(LocalFunctionSymbol localFunc,
-                                    SyntaxNode syntax,
-                                    bool isWrite)
-        {
-            LocalFuncUsages usages = GetOrCreateLocalFuncUsages(localFunc);
-            var state = isWrite ? usages.WrittenVars : usages.ReadVars;
-
-            // Start at slot 1 (slot 0 just indicates reachability)
-            for (int slot = 1; slot < state.Capacity; slot++)
-            {
-                if (state[slot])
-                {
-                    if (isWrite)
-                    {
-                        SetSlotAssigned(slot);
-                    }
-                    else
-                    {
-                        var symbol = variableBySlot[slot].Symbol;
-                        CheckAssigned(symbol, syntax, slot);
-                    }
-                }
+                UnionWith(ref this.State, ref usages.WrittenVars);
             }
 
             usages.LocalFuncVisited = true;
+        }
+
+        private void ReplayReads(ref BitVector reads, SyntaxNode syntax)
+        {
+            // Start at slot 1 (slot 0 just indicates reachability)
+            for (int slot = 1; slot < reads.Capacity; slot++)
+            {
+                if (reads[slot])
+                {
+                    var symbol = variableBySlot[slot].Symbol;
+                    CheckAssigned(symbol, syntax, slot);
+                }
+            }
         }
 
         private int RootSlot(int slot)
@@ -92,7 +85,8 @@ namespace Microsoft.CodeAnalysis.CSharp
         public override BoundNode VisitLocalFunctionStatement(BoundLocalFunctionStatement localFunc)
         {
             var oldMethodOrLambda = this.currentMethodOrLambda;
-            this.currentMethodOrLambda = localFunc.Symbol;
+            var localFuncSymbol = localFunc.Symbol;
+            this.currentMethodOrLambda = localFuncSymbol;
 
             var oldPending = SavePending(); // we do not support branches into a lambda
 
@@ -101,17 +95,17 @@ namespace Microsoft.CodeAnalysis.CSharp
             var savedState = this.State;
             this.State = this.ReachableState();
 
-            var usages = GetOrCreateLocalFuncUsages(localFunc.Symbol);
+            var usages = GetOrCreateLocalFuncUsages(localFuncSymbol);
             var oldReads = usages.ReadVars;
             usages.ReadVars = BitVector.Empty;
 
-            if (!localFunc.WasCompilerGenerated) EnterParameters(localFunc.Symbol.Parameters);
+            if (!localFunc.WasCompilerGenerated) EnterParameters(localFuncSymbol.Parameters);
 
             var oldPending2 = SavePending();
 
             // If this is an iterator, there's an implicit branch before the first statement
             // of the function where the enumerable is returned.
-            if (localFunc.Symbol.IsIterator)
+            if (localFuncSymbol.IsIterator)
             {
                 PendingBranches.Add(new PendingBranch(null, this.State));
             }
@@ -123,19 +117,19 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             Location location = null;
 
-            if (!localFunc.Symbol.Locations.IsDefaultOrEmpty)
+            if (!localFuncSymbol.Locations.IsDefaultOrEmpty)
             {
-                location = localFunc.Symbol.Locations[0];
+                location = localFuncSymbol.Locations[0];
             }
 
-            LeaveParameters(localFunc.Symbol.Parameters, localFunc.Syntax, location);
+            LeaveParameters(localFuncSymbol.Parameters, localFunc.Syntax, location);
 
             LocalState stateAtReturn = this.State;
             foreach (PendingBranch pending in pendingReturns)
             {
                 this.State = pending.State;
                 BoundNode branch = pending.Branch;
-                LeaveParameters(localFunc.Symbol.Parameters, branch?.Syntax,
+                LeaveParameters(localFuncSymbol.Parameters, branch?.Syntax,
                                 branch?.WasCompilerGenerated == true
                                     ? location : null);
                 IntersectWith(ref stateAtReturn, ref this.State);
@@ -143,7 +137,7 @@ namespace Microsoft.CodeAnalysis.CSharp
 
             // Check for changes to the read and write sets
             if (RecordChangedVars(ref usages.WrittenVars,
-                                  ref stateAtReturn.Assigned,
+                                  ref stateAtReturn,
                                   ref oldReads,
                                   ref usages.ReadVars) &&
                 usages.LocalFuncVisited)
@@ -168,12 +162,13 @@ namespace Microsoft.CodeAnalysis.CSharp
             usages.ReadVars[slot] = true;
         }
 
-        private bool RecordChangedVars(ref BitVector oldWrites,
-                                       ref BitVector newWrites,
+        private bool RecordChangedVars(ref LocalState oldWrites,
+                                       ref LocalState newWrites,
                                        ref BitVector oldReads,
                                        ref BitVector newReads)
         {
-            bool anyChanged = RecordCapturedChanges(ref oldWrites, ref newWrites);
+            bool anyChanged = IntersectWith(ref oldWrites, ref newWrites);
+
             anyChanged |= RecordCapturedChanges(ref oldReads, ref newReads);
 
             return anyChanged;
@@ -229,8 +224,7 @@ namespace Microsoft.CodeAnalysis.CSharp
             LocalFuncUsages usages;
             if (!_localFuncVarUsages.TryGetValue(localFunc, out usages))
             {
-                usages = new LocalFuncUsages();
-                _localFuncVarUsages[localFunc] = usages;
+                usages = _localFuncVarUsages[localFunc] = new LocalFuncUsages(UnreachableState());
             }
             return usages;
         }

--- a/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
+++ b/src/Compilers/CSharp/Portable/FlowAnalysis/DataFlowPass.cs
@@ -2277,7 +2277,8 @@ namespace Microsoft.CodeAnalysis.CSharp
                 Normalize(ref other);
             }
 
-            if (other.Assigned[0]) self.Assigned[0] = true;
+            if (!other.Reachable) self.Assigned[0] = true;
+
             for (int slot = 1; slot < self.Assigned.Capacity; slot++)
             {
                 if (other.Assigned[slot] && !self.Assigned[slot])

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
@@ -8,6 +8,7 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
     public class LocalFunctions : FlowTestBase
     {
         [Fact]
+        [WorkItem(14046, "https://github.com/dotnet/roslyn/issues/14046")]
         public void UnreachableAfterThrow()
         {
             var comp = CreateCompilationWithMscorlib(@"
@@ -35,6 +36,7 @@ class C
         }
 
         [Fact]
+        [WorkItem(13739, "https://github.com/dotnet/roslyn/issues/13739")]
         public void UnreachableRecursion()
         {
             var comp = CreateCompilationWithMscorlib(@"
@@ -81,6 +83,7 @@ class C
         }
 
         [Fact]
+        [WorkItem(13739, "https://github.com/dotnet/roslyn/issues/13739")]
         public void MutualRecursiveUnreachable()
         {
             var comp = CreateCompilationWithMscorlib(@"
@@ -112,6 +115,7 @@ class C
         }
 
         [Fact]
+        [WorkItem(13739, "https://github.com/dotnet/roslyn/issues/13739")]
         public void AssignedInDeadBranch()
         {
             var comp = CreateCompilationWithMscorlib(@"

--- a/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
+++ b/src/Compilers/CSharp/Test/Semantic/FlowAnalysis/LocalFunctions.cs
@@ -8,6 +8,141 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests
     public class LocalFunctions : FlowTestBase
     {
         [Fact]
+        public void UnreachableAfterThrow()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class C
+{
+  public void M3()
+  {
+    int x, y;
+    void Local()
+    {
+      throw null;
+      x = 5; // unreachable code
+      System.Console.WriteLine(y);
+    }
+
+    Local();
+    System.Console.WriteLine(x);
+    System.Console.WriteLine(y);
+  }
+}");
+            comp.VerifyDiagnostics(
+                // (10,7): warning CS0162: Unreachable code detected
+                //       x = 5; // unreachable code
+                Diagnostic(ErrorCode.WRN_UnreachableCode, "x").WithLocation(10, 7));
+        }
+
+        [Fact]
+        public void UnreachableRecursion()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class C
+{
+    public void M()
+    {
+        int x, y;
+        void Local()
+        {
+            Local();
+            x = 0;
+        }
+        Local();
+        x++;
+        y++;
+    }
+}");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void ReadBeforeUnreachable()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class C
+{
+    public void M()
+    {
+        int x;
+        void Local()
+        {
+            x++;
+            Local();
+        }
+        Local();
+        x++;
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (12,9): error CS0165: Use of unassigned local variable 'x'
+                //         Local();
+                Diagnostic(ErrorCode.ERR_UseDefViolation, "Local()").WithArguments("x").WithLocation(12, 9));
+        }
+
+        [Fact]
+        public void MutualRecursiveUnreachable()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+class C
+{
+    public static void M()
+    {
+        int x, y, z;
+        
+        void L1()
+        {
+            L2();
+            x++; // Unreachable
+        } 
+        void L2()
+        {
+            L1();
+            y++; // Unreachable
+        }
+
+        L1();
+        // Unreachable code, so everything should be definitely assigned
+        x++;
+        y++;
+        z++;
+    }
+}");
+            comp.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void AssignedInDeadBranch()
+        {
+            var comp = CreateCompilationWithMscorlib(@"
+using System;
+
+class Program
+{
+    public static void Main(string[] args)
+    {
+        int u;
+        M();
+    https://github.com/dotnet/roslyn/issues/13739
+        Console.WriteLine(u); // error: use of unassigned local variable 'u'
+        return;
+
+        void M()
+        {
+            goto La;
+        Lb: return;
+        La: u = 3;
+            goto Lb;
+        }
+    }
+}");
+            comp.VerifyDiagnostics(
+                // (10,5): warning CS0164: This label has not been referenced
+                //     https://github.com/dotnet/roslyn/issues/13739
+                Diagnostic(ErrorCode.WRN_UnreferencedLabel, "https").WithLocation(10, 5));
+        }
+
+        [Fact]
         public void InvalidBranchOutOfLocalFunc()
         {
             var comp = CreateCompilationWithMscorlib(@"

